File: src/main/java/org/jetbrains/bsp/bazel/server/bazel/BazelRunnerCommandBuilder.java
package org.jetbrains.bsp.bazel.server.bazel;

public class BazelRunnerCommandBuilder {

  private static final String BAZEL_AQUERY_COMMAND = "aquery";
  private static final String BAZEL_BUILD_COMMAND = "build";
  private static final String BAZEL_CLEAN_COMMAND = "clean";
  private static final String BAZEL_FETCH_COMMAND = "fetch";
  private static final String BAZEL_INFO_COMMAND = "info";
  private static final String BAZEL_QUERY_COMMAND = "query";
  private static final String BAZEL_RUN_COMMAND = "run";
  private static final String BAZEL_TEST_COMMAND = "test";

  private final BazelRunner bazelRunner;

  BazelRunnerCommandBuilder(BazelRunner bazelRunner) {
    this.bazelRunner = bazelRunner;
  }

  public BazelRunnerBuilder aquery() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_AQUERY_COMMAND);
  }

  public BazelRunnerBuilder build() {
    return new BazelRunnerBuilderWithoutMnemonics(bazelRunner, BAZEL_BUILD_COMMAND);
  }

  public BazelRunnerBuilder clean() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_CLEAN_COMMAND);
  }

  public BazelRunnerBuilder fetch() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_FETCH_COMMAND);
  }

  public BazelRunnerBuilder info() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_INFO_COMMAND);
  }

  public BazelRunnerBuilder run() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_RUN_COMMAND);
  }

  public BazelRunnerBuilder query() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_QUERY_COMMAND);
  }

  public BazelRunnerBuilder test() {
    return new BazelRunnerBuilder(bazelRunner, BAZEL_TEST_COMMAND);
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/BazelRunnerCommandBuilder.kt
package org.jetbrains.bsp.bazel.bazelrunner

class BazelRunnerCommandBuilder internal constructor(private val bazelRunner: BazelRunner) {
  fun aquery() = BazelRunnerBuilder(bazelRunner, listOf("aquery"))
  fun clean()  = BazelRunnerBuilder(bazelRunner, listOf("clean"))
  fun fetch()  = BazelRunnerBuilder(bazelRunner, listOf("fetch"))
  fun info()   = BazelRunnerBuilder(bazelRunner, listOf("info"))
  fun run()    = BazelRunnerBuilder(bazelRunner, listOf("run")).withUseBuildFlags()
  fun mod(subcommand: String) = BazelRunnerBuilder(bazelRunner, listOf("mod", subcommand))
  fun graph() = mod("graph")
  fun deps() = mod("deps")
  fun allPaths() = mod("all_paths")
  fun path() = mod("path")
  fun explain() = mod("explain")
  fun showRepo() = mod("show_repo")
  fun showExtension() = mod("show_extension")
  fun query()  = BazelRunnerBuilder(bazelRunner, listOf("query"))
  fun cquery()  = BazelRunnerBuilder(bazelRunner, listOf("cquery"))

  fun build()  = BazelRunnerBuildBuilder(bazelRunner, listOf("build")).withUseBuildFlags()
  fun mobileInstall() = BazelRunnerBuildBuilder(bazelRunner, listOf("mobile-install")).withUseBuildFlags()
  fun test()   = BazelRunnerBuildBuilder(bazelRunner, listOf("test")).withUseBuildFlags()
  fun coverage() = BazelRunnerBuildBuilder(bazelRunner, listOf("coverage")).withUseBuildFlags()
}
